app/models/spree/order_decorator.rb

```diff
@@ -17,6 +17,7 @@
 
   validates :customer, presence: true, if: :require_customer?
   validate :products_available_from_new_distribution, :if => lambda { distributor_id_changed? || order_cycle_id_changed? }
+  validate :disallow_guest_order, if: lambda { using_guest_checkout? && registered_email? }
```

Comment:
I would move the logic into `disallow_guest_order`. Much easier to read and everything is in one place. Still nice to keep the separate methods below though.